src/Visualization/Python/ApplyPointwise.py

```diff
@@ -63,49 +76,176 @@ def get_tensor_component_names(tensor_name: str,
+        return get_tensor_component_names(self.dataset_name, self.tensor_type)
+
+    def get(self, all_tensor_data: Dict[str, Tensor],
+            element: Optional[Element]):
+        tensor_data = all_tensor_data[self.dataset_name]
+        if element:
+            components = np.asarray(tensor_data)[:, element.data_slice]
+            return self.tensor_type(components)
+        else:
+            return tensor_data
+
+
+@dataclass(frozen=True)
+class ElementArg(KernelArg):
+    """Kernel argument retrieved from an element in the computational domain"""
+    element_attr: str
+
+    def get(self, all_tensor_data: Dict[str, Tensor], element: Element):
+        return getattr(element, self.element_attr)
+
+
+def parse_kernel_arg(arg: inspect.Parameter,
+                     map_input_names: Dict[str, str]) -> KernelArg:
+    """Determine how data for a kernel function argument will be retrieved
+
+    Examines the name and type annotation of the argument. The following
+    arguments are supported:
+
+    - Mesh: Annotate the argument with a 'Mesh' type.
+    - Coordinates: Annotate the argument with a 'tnsr.I' type and name it
+        "logical_coords" / "logical_coordinates" / "xi" or "inertial_coords" /
+        "inertial_coordinates" / "x".
+    - Jacobians: Annotate the argument with a 'Jacobian' or 'InverseJacobian'.
+    - Any tensor dataset: Annotate the argument with the tensor type. By
+        default, the argument name is transformed to CamelCase to determine the
+        dataset name in the volume data file. Specify a mapping in
+        'map_input_names' to override the default.
+
+    For example, the following function is a possible kernel:
+
+        def one_index_constraint(
+            psi: Scalar[DataVector],
+            mesh: Mesh[3],
+            inv_jacobian: InverseJacobian[DataVector, 3],
+        ) -> tnsr.i[DataVector, 3]:
+            # ...
+
+    The function can also be a binding of a C++ function, such as this:
+
+        tnsr::i<DataVector, 3> one_index_constraint(
+            const Scalar<DataVector>& psi,
+            const Mesh<3>& mesh,
+            const InverseJacobian<
+              DataVector, 3, Frame::ElementLogical, Frame::Inertial>&
+              inv_jacobian) {
+            // ...
+        }
+
+    Arguments:
+      arg: The function argument. Must have a type annotation.
+      map_input_names: A map of argument names to dataset names (optional).
+        By default the argument name is transformed to CamelCase.
+
+    Returns: A 'KernelArg' object that knows how to retrieve the data for the
+      argument.
+    """
+    if arg.annotation is inspect.Parameter.empty:
+        raise ValueError(
+            "Please annotate all arguments in the kernel function. "
+            f"Argument '{arg.name}' has no annotation.")
+
+    if arg.annotation in Mesh.values():
+        return ElementArg("mesh")
+    elif (arg.name in ["logical_coords", "logical_coordinates", "xi"]
```

Comment:
Also removed the "xi" option because maybe that can be used to mean something other than logical coords (some people name the displacement field in elasticity "xi").